src/construction.cpp

```diff
@@ -2550,42 +2578,82 @@ build_reqs get_build_reqs_for_furn_ter_ids(
             total_builds[build.id] = 0;
         }
         total_builds[build.id] += count;
-        std::string build_pre_ter = build.pre_terrain;
+        if( build.pre_terrain.size() > 1 ) {
+            debugmsg( "get_build_reqs_for_furn_ter_ids tried to get reqs for %s which has multiple pre_terrain",
+                      build.str_id.str() );
+            return;
+        }
+        std::string build_pre_ter = build.pre_terrain.empty() ? "" : *( build.pre_terrain.begin() );
         while( !build_pre_ter.empty() ) {
-            for( const construction &pre_build : constructions ) {
-                if( pre_build.category == construction_category_REPAIR ) {
-                    continue;
-                }
-                if( ( pre_build.post_terrain.empty() ||
-                      ( !pre_build.post_is_furniture &&
-                        ter_id( pre_build.post_terrain ) != base_ter ) ) &&
-                    ( pre_build.pre_terrain.empty() ||
-                      ( pre_build.post_is_furniture &&
-                        ter_id( pre_build.pre_terrain ) == base_ter ) ) &&
-                    pre_build.post_terrain == build_pre_ter &&
-                    pre_build.pre_terrain != build.post_terrain ) {
-                    if( total_builds.find( pre_build.id ) == total_builds.end() ) {
-                        total_builds[pre_build.id] = 0;
+            bool found_pre = false;
+            // only consider DECORATE constructions if there's no other way to build the target
+            // this will allow painting walls, but will skip un-painting walls as a way to make a wall
+            for( bool allow_decorate : {
+                     false, true
+                 } ) {
+                for( const construction &pre_build : constructions ) {
+                    if( ( pre_build.category == construction_category_DECORATE ) != allow_decorate ) {
+                        continue;
+                    }
```

Comment:
The old code checks DECORATE constructions with equal priority to non-DECORATE, leading to, e.g., choosing "take paint off wall" as the definitive way to build a wall.